### ISSUE TITLE
An unhandled exception while Generating ES5 bundles

### ISSUE THREAD

**Issue by ronfogel** [angular/angular-cli#15493]
# 🐞 Bug report

### Command (mark with an `x`)
```
- [ ] new
- [x ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Is this a regression?
Yes, the previous version in which this bug was not present was 8.2.1, maybe 8.2.2

### Description
When using ng build in circle ci, it throws an error (screenshot attached):
```
Generating ES5 bundles for differential loading...
An unhandled exception occurred: cancel after 1 retries!
```
<img width="1177" alt="Screen Shot 2019-08-31 at 12 42 29" src="https://user-images.githubusercontent.com/16741829/64062240-e79c2080-cbec-11e9-979f-1f1bc8024cb0.png">

## 🔬 Minimal Reproduction
Unfortunately, I cant provide a way to reproduce it as it happens only in CI and I am unable to get the logs. I think it's related to the size of my app but I cant be sure.

## 🔥 Exception or Error
<pre><code>
WARNING in budgets, maximum exceeded for initial. Budget 2 MB was exceeded by 2.2 MB.
Generating ES5 bundles for differential loading...
An unhandled exception occurred: cancel after 1 retries!
See "/tmp/ng-8Tr7C1/angular-errors.log" for further details.
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! frontend-mallabee@2.0.1 build-dev: `ng build --configuration dev`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the frontend-mallabee@2.0.1 build-dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/circleci/.npm/_logs/2019-08-31T09_20_07_881Z-debug.log
Exited with code 1
</code></pre>

## 🌍 Your Environment
<pre><code>
@angular-devkit/architect         0.803.2
@angular-devkit/build-angular     0.803.2
@angular-devkit/build-optimizer   0.803.2
@angular-devkit/build-webpack     0.803.2
@angular-devkit/core              8.3.2
@angular-devkit/schematics        8.3.2
@angular/cdk                      8.1.4
@angular/cli                      8.3.2
@angular/flex-layout              8.0.0-beta.27
@angular/material                 8.1.4
@ngtools/webpack                  8.3.2
@schematics/angular               8.3.2
@schematics/update                0.803.2
rxjs                              6.5.2
typescript                        3.5.3
webpack                           4.39.2
</code></pre>

**Comment by xawill** [comment]
Same issue here.

**Comment by alan-agius4** [comment]
Hi @ronfogel and @xawill, can anyone of you provide the logs please and take it from there?

Most likely though we'd need a minimal repro.

You can read [here](https://github.com/angular/angular-cli/blob/master/CONTRIBUTING.md#-submitting-an-issue) why this is needed. A good way to make a minimal repro is to create a new app via `ng new repro-app` and adding the minimum possible code to show the problem. Then you can push this repository to github and link it here.

**Comment by xawill** [comment]
Dear @alan-agius4,

Thank you very much for getting back to us so quickly. Here are my logs :

>
\> OBFUSCATED_PROJECT@0.0.0 ng OBFUSCATED_CI_PATH
\> ng "build"
chunk {main} main-es2015.js, main-es2015.js.map (main) 8.59 MB [initial] [rendered]
chunk {polyfills} polyfills-es2015.js, polyfills-es2015.js.map (polyfills) 264 kB [initial] [rendered]
chunk {polyfills-es5} polyfills-es5-es2015.js, polyfills-es5-es2015.js.map (polyfills-es5) 584 kB [initial] [rendered]
chunk {runtime} runtime-es2015.js, runtime-es2015.js.map (runtime) 6.16 kB [entry] [rendered]
chunk {styles} styles-es2015.js, styles-es2015.js.map (styles) 2.09 MB [initial] [rendered]
chunk {vendor} vendor-es2015.js, vendor-es2015.js.map (vendor) 10 MB [initial] [rendered]
Date: 2019-08-27T13:16:27.208Z - Hash: cdbe70565be644ecc800 - Time: 62573ms
Generating ES5 bundles for differential loading...
An unhandled exception occurred: cancel after 1 retries!
See "/tmp/ng-EB7H6h/angular-errors.log" for further details.
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! project-creator@0.0.0 ng: `ng "build"`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the OBFUSCATED_PROJECT@0.0.0 ng script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

Here is what is contained in `/tmp/ng-EB7H6h/angular-errors.log` :
> [error] ProcessTerminatedError: cancel after 1 retries!
    at Farm.<anonymous> (OBFUSCATED_CI_PATH/node_modules/worker-farm/lib/farm.js:88:25)
    at Array.forEach (<anonymous>)
    at Farm.<anonymous> (OBFUSCATED_CI_PATH/node_modules/worker-farm/lib/farm.js:82:36)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)

As a programmer, I totally understand why you need a way to reproduce the problem in order to solve it. Unfortunately I cannot, as it only happens when building in the CI (i.e. it doesn't happen locally even though the setup is identical) and of course I cannot give you access to my CI...

I have no clue how to help you help me for this one... But tell me if you have any idea.

**Comment by alan-agius4** [comment]
Seeing the full stacktrace it might be a resource problem since CIRCLE CI will have limited resources If you are using the free plan the resources are 2CPU and 4096MB.

**Comment by xawill** [comment]
It's Atlassian bamboo CI, for me, but I also arrived at the conclusion that it was ressources related. I'll try to tweak the performance parameters of Bamboo.
Any particular reason why 8.3.x is more ressource-hungry than <8.3 ?

**Comment by alan-agius4** [comment]
In 8.3 we are using workers farms which will spawn multiple processes at once to parallelize ES5 bundles generation.

**Comment by ronfogel** [comment]
@alan-agius4 Maybe a good way to maintain it (sounds like a good way to improve performance locally) and provide a solution to CI environments with limited resources would be adding an option to angular.json or some other place to determine whether it should run multiple processes to parallelize ES5 bundles generation or not.

**Comment by pastcompute** [comment]
This has just started happening to me after `Generating ES5 bundles for differential loading...` on a Mac Mini with 2.6GHZ i5 and 16GB of RAM latest MacOS 10.14.6 when I build with `--debug` through Ionic4 (not a CI system), or rather, `ng build` with no options.

Upgraded from 0.803.2 to 0.803.3 etc and still happens

(I'm not presently in a position to make a minimal repo, sorry)

**Comment by pastcompute** [comment]
I was able to fix my build by for the moment changing the `tsconfig.app.json` `target` from `es2015` back down to `es5`, this disabled differential loading.

See https://angular.io/guide/deployment#differential-loading

**Comment by clydin** [comment]
@pastcompute Can you provide the output just prior to the `Generating ES5 bundles for differential loading...`?  Or more specifically, the sizes of the source map files?  Sourcemap processing on node is unfortunately incredibly memory intensive.  We've added an alternative source map processing path when the code size is beyond a certain point but the heuristics may need to be adjusted.  Any information as to file size or number of source entries within the source maps would be helpful in making improvements in this area.

**Comment by daniel-luna-gonzalez** [comment]
> I was able to fix my build by for the moment changing the `tsconfig.app.json` `target` from `es2015` back down to `es5`, this disabled differential loading.
> 
> See https://angular.io/guide/deployment#differential-loading

Thanks it works for me

**Comment by josh-m-sharpe** [comment]
@clydin here ya go:

From CircleCI (failing build):
https://gist.github.com/josh-m-sharpe/864bcf6386e467f1f3218829105ff9b3

And filesizes after compiling (successfully) locally:
https://gist.github.com/josh-m-sharpe/92fd81d1f8cef6e14ca248e6466383c0

**Comment by vamcs** [comment]
I can reproduce the problem by running `ng build` locally and I've tried this with CLI versions: `8.3.2`, `8.2.2`, `8.2.1` and `8.1.1`. Our app is exceeding the budget limits for the production builds and I have been trying to investigate with `webpack-bundle-analyzer` what can be removed, so that's why I needed to run `ng build --stats-json` locally. Right now, the only way this works is by changing the `target` from `es2015` to `es5` in `tsconfig.json` as @pastcompute suggested.

I'm running WSL on Windows.

**Comment by PillowPillow** [comment]
Same issue here with CircleCI. Targeting es5 instead of es2015 works but it shouldn't be a viable solution.

**Comment by dhair-seva** [comment]
Yeah, definitely not a viable solution. I changed it to es5 too. I'm planning to lower my bundle size to see if it fixes it, but that can't happen for a while. I wish this could be resolved without a bundle size change.

**Comment by robert-king** [comment]
I got this bug too, not sure what from, perhaps it's because I did an ng update with --force last week.

switching to es5 target fixed it but don't wanna do that i guess?

here's my package.json in case that helps 

```

{
  "name": "front-end",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start-cookie-monster": "node ../api-tests/bin/www",
    "start": "ng serve",
    "build": "ng build --configuration=production",
    "build:docker": "docker image build -t mybarker-front-end-image .",
    "start:docker": "docker run -p 3000:80 --rm mybarker-front-end-image",
    "test": "ng test",
    "test-fast": "npm run test -- --no-watch --no-progress --browsers=ChromeHeadlessCI",
    "lint": "ng lint",
    "e2e": "ng e2e --port=4100",
    "precommit-check": "echo \"does nothing\"",
    "ng-swag": "node_modules/.bin/ng-swagger-gen -i https://api.dev.mybarker.dev-sites.co.nz/swagger/v1/swagger.json -o src/app/swag-api",
    "open-chrome": "open -a \"Google Chrome\" --args --disable-web-security --user-data-dir"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "~8.2.5",
    "@angular/cdk": "~8.2.0",
    "@angular/common": "~8.2.5",
    "@angular/compiler": "~8.2.5",
    "@angular/core": "~8.2.5",
    "@angular/forms": "~8.2.5",
    "@angular/material": "^8.1.4",
    "@angular/platform-browser": "~8.2.5",
    "@angular/platform-browser-dynamic": "~8.2.5",
    "@angular/router": "~8.2.5",
    "@material-extended/mde": "^2.3.1",
    "core-js": "^3.2.1",
    "hammerjs": "^2.0.8",
    "ngx-material-timepicker": "^5.0.0",
    "ngx-timeago": "^1.0.2",
    "rxjs": "~6.5.3",
    "tslib": "^1.10.0",
    "zone.js": "~0.10.2"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "^0.803.3",
    "@angular/cli": "~8.3.3",
    "@angular/compiler-cli": "~8.2.5",
    "@angular/language-service": "~8.2.5",
    "@types/jasmine": "~3.4.0",
    "@types/jasminewd2": "~2.0.3",
    "@types/node": "~12.7.4",
    "codelyzer": "^5.1.0",
    "jasmine-core": "~3.4.0",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~4.3.0",
    "karma-chrome-launcher": "~3.1.0",
    "karma-coverage-istanbul-reporter": "~2.1.0",
    "karma-jasmine": "~2.0.1",
    "karma-jasmine-html-reporter": "^1.4.2",
    "ng-swagger-gen": "^1.8.1",
    "protractor": "~5.4.0",
    "shelljs": "^0.8.3",
    "ts-node": "~8.3.0",
    "tslint": "~5.20.0",
    "typescript": "^3.5.3"
  }
}

```

**Comment by jonstelly** [comment]
Bumping my build agents from 2GB to 8GB of RAM fixed this for me.  I'm not sure if 4 or 6GB would have worked, didn't have time to spend on investigation today.

**Comment by robert-king** [comment]
My macbook pro is 8GB RAM but still got the error

**Comment by akvaliya** [comment]
Same issue here with 4GB RAM in linux server.

**Comment by josh-m-sharpe** [comment]
@vikerman is it possible to test your change by installing @angular/cli from master somehow?

**Comment by clydin** [comment]
Snapshot releases for the relevant package can be found here: https://github.com/angular/angular-devkit-build-angular-builds


All the CLI packages have equivalent snapshot repositories as well.

Please note that these should generally only be used for testing as they represent a version in active development.

**Comment by dimamarksman** [comment]
1. `ng build` renders a bundle report where is `polyfills-es5-es2015.[hash].js` but if we open a `dist` folder there is no polyfills-es5-es2015.[hash].js file.
2. bundle sizes in the report don't match actual file sizes on the disk, it leads to the `budgets` issue.

**Comment by xawill** [comment]
The number of affected users seems to be quite high, in the end. Any news from the dev team ? Have you been able to precisely identify the issue ?

**Comment by josh-m-sharpe** [comment]
This seems to be resolved for me using the latest CLI.

**Comment by csegyud** [comment]
I have updated to the latest Angular 8 (8.2.8) and CLI (8.3.6) versions with `compilerOptions / target:'es2015'` but unfortunately I still get the same heap error. 
I have to set target back to `es5` to make it work.  

Although, I have tried it with my existing project and not a new one. I will check if the new CLI changed anything in the config files I need to adjust in my existing config.

**Comment by clydin** [comment]
For this issue, the `@angular-devkit/build-angular` package is the critical element.  Ensure that it is at the latest version (currently `0.803.6`).

Please also try to keep this issue centered on the originating issue of memory usage.  Other issues should be tracked separately and may already be in the process of being fixed (#15666).

**Comment by josh-m-sharpe** [comment]
I take back my previous comment that this has been fixed.  I am still seeing the issue using these packages:

```
    "@angular/common": "~8.2.8",
    "@angular/core": "~8.2.8",
    "@angular/forms": "~8.2.8",
    "@angular/platform-browser": "~8.2.8",
    "@angular/platform-browser-dynamic": "~8.2.8",
    "@angular/router": "~8.2.8",
    "@angular-devkit/architect": "^0.803.6",
    "@angular-devkit/build-angular": "^0.803.6",
    "@angular-devkit/core": "^8.3.6",
    "@angular-devkit/schematics": "^8.3.6",
    "@angular/cli": "^8.3.6",
    "@angular/compiler": "^8.2.8",
    "@angular/compiler-cli": "^8.2.8",
    "@angular/language-service": "^8.2.8",
```

**Comment by andrexx** [comment]
Got the same error with the local build inside docker container.
Updating packages didn't help. Switching target to es5 solve issue.

**Comment by JDeVri01** [comment]
Same issue with the worker-related exception only occurring when using CI (AWS CodeBuild) versus manually building on my local computer. Bumped the CodeBuild environment from 3GB RAM/2vCPU to 7GB RAM/4vCPU, and that resolved the error.

**Comment by petersandor** [comment]
Experiencing the same issue. Only solution is to not generate bundles for differential loading as described above ☝️https://github.com/angular/angular-cli/issues/15493#issuecomment-529098088

**Comment by krzysiek3d** [comment]
Try with this: 
node_modules/@angular/cli/bin/ng build --prod
or:
node --max_old_space_size=8192 node_modules/@angular/cli/bin/ng build --prod

**Comment by jimmykane** [comment]
Same here: 

```
An unhandled exception occurred: Call retries were exceeded
See "/private/var/folders/ms/qq0y5v_d0_q8q1sk43c87b080000gn/T/ng-xZOLD0/angular-errors.log" for further details.

event": "Allocation failed - JavaScript heap out of memory",
    "trigger": "FatalError",
```

**Comment by jimmykane** [comment]
If it helps this solved it for me: 

```
"build-beta": "NODE_OPTIONS=--max-old-space-size=8192 ng build --verbose --configuration beta",
``` 

prepending to the scripts part of your app build the 
```
NODE_OPTIONS=--max-old-space-size=8192
```

Should increase memory size for those that this can help them

**Comment by clydin** [comment]
For anyone experiencing the memory issue, please ensure that the CLI is on the latest version (currently 8.3.7).  Also, please ensure that the `@angular-devkit/build-angular` package is also at the latest (currently 0.803.7).  There are also benefits to using the current version of Node.js (12.11) if possible.

For the team to provide more assistance and troubleshoot these occurrences, please provide additional information regarding your projects.  An example repository that can consistently reproduce the behavior would be ideal.  However, the settings used during the build are paramount (especially optimization, aot, and sourcemaps).  An output of a working ES5 only build with size information would also be beneficial.

**Comment by jimmykane** [comment]
@clydin here is a repo https://github.com/jimmykane/quantified-self/ you can try to do a beta build and it will fail

**Comment by richarddavenport** [comment]
Upgraded our app from 7 to 8 and ran into this today. @jimmykane solved my issue by increasing the max old space size flag. 

`NODE_OPTIONS=--max-old-space-size=8192 ng build --prod`

Our build right now is pretty large. The largest file that is built is the `main-es5.xxxx.js` file at 6.2MB.

**Comment by csegyud** [comment]
@richarddavenport , I can also confirm that the NODE_OPTIONS solution from @jimmykane is working, and in appr. 3 minutes it builds without errors. Peak memory usage exceeds 4 GB. Not too fast though on an Core i7 / 32GB / SSD machine.

And I can also see that my es5 bundle is 14MB while the es2015 bundle is 17MB. Although the polyfill for es2015 is 460KB less than es5 polyfill, the bundle size is much much more than I can save on the polyfill files. So I have a feeling that it is not mature yet - while the idea sounds really good.

**Comment by clydin** [comment]
@csegyud Are those numbers the on-disk sizes?  The ES5 bundles are essentially the ES2015 bundles with additional code structures added to mimic ES2015 features in ES5.

**Comment by mogwai1** [comment]
Faced with same issues after migration from Angular 7 to 8.
1. error with CI build: "javascript heap out of memory"
Fixed via --max-old-space-size=4096
2. error with local build inside Docker for Windows container: "Call retries were exceeded"
Fixed via allocating 4Gb of memory to Docker engine (it was 2Gb by default in docker settings).

**Comment by csegyud** [comment]
@clydin, These are on disk sizes and result of a production build. 

The sizes of *.js in my dist folder:
```
17720682 bytes main-es2015.c54c6ee19c4c250f0f9a.js
14716590 bytes main-es5.c54c6ee19c4c250f0f9a.js
  132733 bytes polyfills-es2015.833fec98dca74639f6a6.js
  605953 bytes polyfills-es5.12a64184cfbc85c720a9.js
    6263 bytes runtime-es2015.d6c52737d4587c65265f.js
    5606 bytes runtime-es5.d6c52737d4587c65265f.js
```

**Comment by clydin** [comment]
Are you sure the project has optimizations enabled? The polyfill sizes for a new application in production are as follows:
```
37,293   polyfills-es2015.f80ea2f8856ba3d344c6.js
126,094  polyfills-es5.1323fe914a6541b59b65.js
```
And for a development (with AOT) build:
```
132,713  polyfills-es2015.js
636,519  polyfills-es5.js
```

Regardless, the issue appears to be triggered by having a javascript file that is in the 10MB+ range.  Work is ongoing to remedy the situation and the team will update as progress is made.

**Comment by csegyud** [comment]
@clydin, sure, you were right, the optimization was turned off. Turning it on I get 5MB bundles for both versions (es2015 is smaller by a 200 KB) which is correct enough. BTW, the compilation took ~5:30 minutes now and the memory usage peaked almost 6 GB. :) No problem though, it just fits the budget (32GB). Thanks for the hint.

**Comment by jimmykane** [comment]
Is there a version that we can rollback to that does not have this issue? 

The build procedure for a 1MB app has became a nightmare.

**Comment by egorkel-da14** [comment]
@jimmykane Reverting to these versions helps me
```
"@angular-devkit/build-angular": "~0.802.2",
"@angular/cli": "~8.2.2"
```

**Comment by jimmykane** [comment]
@egorkel-da14 thanks so much. Tired of waiting 30+ mins to build an app.

**Comment by clydin** [comment]
Differential loading was introduced in version 8.0.0.  Reverting to 8.2.0 will cause production builds to be longer not shorter as two full builds of the application will be performed if differential loading is enabled.  This will be even more pronounced on Windows systems where file access is more costly.

Differential loading is controlled by both the target option within the application’s tsconfig and the browserslist file.  The browserslist file should ideally be tailored to the application’s browser support requirements.  This allows the CLI to optimize the code/stylesheets and produce the required set of application variants for the application’s target browsers.  If only browsers that support ES modules are required then the ES5 variant will not be created.
If only an ES5 variant is preferred, the tsconfig target can be set to ES5 and only an ES5 variant will be created.

**Comment by egorkel-da14** [comment]
Thanks for your notes, but in my case I need both builds. So 8.2.0 suits me.

**Comment by clydin** [comment]
8.3.x would be the preferred version in that case.  It is faster than 8.2.x.  As an example, an application with 3500 lazy routes (that is not a typo), performs a production build with differential loading using 8.2.2 in ~15.5 minutes.  With 8.3.8, that build only takes ~6.5 minutes.

It is also important to note that the new build pipeline can perform the differential loading processing of each output file in parallel using a combination of separate processes and threads (if using Node.js 12+).  For applications that leverage lazy loading (a recommended best practice), not only will the end user experience improve but build times will improve as well.

**Comment by egorkel-da14** [comment]
The main problem discussed in this thread is that CI systems have limited resources. So parallelism causes build to fail. That's why we revert to prev version. And I know benefits of the new version.

**Comment by clydin** [comment]
That's not why its failing. The parallelism adjusts based the executing system.  It's failing due to the large size of the main files combined with inefficiencies in the AST traverse processing pipeline of one of the CLI's dependencies.  This is in the process of being corrected.

Beyond this, as a best practice, lazy loading should be strongly considered when the size of the main file becomes larger than 2MB.  As this can greatly affect the initial interactivity time for end-users.

**Comment by tbu567** [comment]
> @clydin here ya go:
> 
> From CircleCI (failing build):
> https://gist.github.com/josh-m-sharpe/864bcf6386e467f1f3218829105ff9b3
> 
> And filesizes after compiling (successfully) locally:
> https://gist.github.com/josh-m-sharpe/92fd81d1f8cef6e14ca248e6466383c0

Regarding CircleCI, in my case I pushed the button "Rerun from failed" it solved the problem.

**Comment by leo6104** [comment]
I got same problem with 9.0.0-next.9 version & Gitlab CI

1. In my project, main.js file size is 1.8mb 
2. 9.0.0-next.7 ~ next.9 not work
3. In Gitlab CI, i use Node 10.x latest image
4. tried to disable sourceMap compile  but it doesn't work.
5. In local build, it success. Only problem on Gitlab CI

```
@angular-devkit/architect          0.900.0-next.9
@angular-devkit/build-angular      0.900.0-next.9
@angular-devkit/build-ng-packagr   0.900.0-next.9
@angular-devkit/build-optimizer    0.900.0-next.9
@angular-devkit/build-webpack      0.900.0-next.9
@angular/cli                       9.0.0-next.9
@ngtools/webpack                   9.0.0-next.9
ng-packagr                         5.6.1
typescript                         3.5.3
webpack                            4.41.0
```

**Comment by niojihy** [comment]
Adding this to my package.json solved for me:

```
"browserslist" : [
        "last 2 Chrome versions"
    ],
```

**Comment by miguellira** [comment]
The following `package.json` dependencies along with Node **12.8.1** worked for me.
```
{
    "@angular-devkit/build-angular": "~0.803.10",
    "@angular/cli": "~8.3.10",
}
```
Also, upgraded to Node **12.12.0** and didn't see a regression.

**Comment by vikerman** [comment]
Fixed - But needs verification.

**Comment by lucipacurar** [comment]
```
Angular CLI: 9.0.0-next.12
Node: 10.16.3
OS: darwin x64
Angular: 9.0.0-next.11
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, upgrade

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-next.11
@angular-devkit/build-angular     0.900.0-next.11
@angular-devkit/build-optimizer   0.900.0-next.11
@angular-devkit/build-webpack     0.900.0-next.11
@angular-devkit/core              9.0.0-next.12
@angular-devkit/schematics        9.0.0-next.12
@angular/cdk                      9.0.0-next.0
@angular/cli                      9.0.0-next.12
@ngtools/webpack                  9.0.0-next.11
@schematics/angular               9.0.0-next.12
@schematics/update                0.900.0-next.12
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.1
```

I get the same error in small/medium sized project with the config above.

```
Generating ES5 bundles for differential loading...
An unhandled exception occurred: Call retries were exceeded
```

**Comment by dstansb** [comment]
`ng update @angular/cli` (updated to `8.3.13`) appears to have solved this issue for me

**Comment by parliament718** [comment]
Simply upgrading my Mac Mini Bamboo CI server from Node 10 -> Node 12.10 fixed my builds.

**Comment by vikerman** [comment]
We will spend some time investigating this trying to see if memory usage can be improved.

**Comment by lucipacurar** [comment]
I tried the latest version again: `ng update @angular/cli @angular/core --next` but I still get the same error. The good thing is that now it produces some logs and the only thing I can understand from there is that it throws an out of memory exception, at about 2Gb of memory used.

**Comment by clydin** [comment]
The latest releases of the CLI contain a variety of improvements to mitigate potential memory issues.  However, as Node.js has a default memory limit of 2GB per process, very large projects may still need to increase the memory value for the Node.js process.  This can be accomplished via a command similar too the following:
`node --max_old_space_size=4096 node_modules/@angular/cli/bin/ng build`
Note the `4096 ` value essentially represents a maximum memory usage value for the node process in megabytes.  For this example, that would be 4GB of memory.  Ensure that the system has enough physical memory to support the value chosen.

It is also recommended to use the latest version of Node 12 (currently 12.13.0) as this contains important improvements related to memory usage.

**Comment by neil-119** [comment]
In our case, we are only supporting modern browsers, so what worked for me is modifying the `browserlist` file to only have `Chrome >= 61`. See this SO [answer](https://stackoverflow.com/a/58264797/1860415).

**Comment by paullryan** [comment]
@clydin and @vikerman this is still failing for me with a large project I have 6656 for memory on the node process and my project is currently using v9.0.0-rc.6 so I believe your fixes should be visible yet it's still failing the same way. I've attached one of the stack traces in case that helps.
[Stacktrace gist](https://gist.github.com/paullryan/8a892b9032c178449f5d2aa9491a90db)

**Comment by clydin** [comment]
@paullryan Can you open a new issue describing the problem you are encountering?  If possible please include the full output of `ng version`, the build command used, the relevant sections of the configuration file for the used build command (build options & configuration options), and also the console output of the build command with different loading disabled.  You should be able to use a temporary `browserslist` file with only `Chrome 75` for the later.

**Comment by paullryan** [comment]
@clydin I realized it was because the prod flag was not being utilized so sorry for the the false alarm, thanks for the response but no need for a new issue.

**Comment by planet4us** [comment]
> Adding this to my package.json solved for me:
> 
> ```
> "browserslist" : [
>     ],
> ```

This worked in conjunction with removing the file _**browserslist**_

**Comment by robelcik** [comment]
Hello. It happened to me after upgrading to angular/cli@8.3.21 and angular-devkit/build-angular@0.803.21. Going back to .20 "fixed" the issue.

**Comment by Harika511** [comment]
> Hello. It happened to me after upgrading to angular/cli@8.3.21 and angular-devkit/build-angular@0.803.21. Going back to .20 "fixed" the issue.
How to go back to .20. please clarify me

**Comment by robelcik** [comment]
@Harika511 
In package.json:
```
  "devDependencies": {
    "@angular-devkit/build-angular": "0.803.20",
    "@angular/cli": "8.3.20",
```

Note missing "~" or "^" characters.

**Comment by strogof** [comment]
@robelcik many many thanks

**Comment by robertrandallphillips** [comment]
Definetely started hanging on differential loading for me after I ran npm audit fix yesterday, which updated @angular-devkit/build-angular from ^0.802.2 to ^0.803.21. If I roll back to 0.802.2 I can build again. Even 0.803.0 doesn't work.

**Comment by mokth** [comment]
install latest nodejs solved my problem. no need to change anything :)

**Comment by jaime-alex** [comment]
I use angular 8.2.10, upgrade nodejs to v12.14.0 solved my problem.

**Comment by bdiriamine** [comment]
run cmd 
ionic cordova run android --prod 
should work ..
have a good day ;)

**Comment by ishfaak** [comment]
Hi bdiriamine,

I had so so so much errors and ionic cordova build --release android nothing helped... IONIC Should know (I am using IONIC 5) for the Mobile app to release something hassle free... Your Answer saved me from two days Work.... Thank you so so so much. The Built the APP as you said.

ionic cordova run android --prod

Happy New year Cheers.

Thank you

**Comment by PercentageCorporation** [comment]
> @Harika511
> In package.json:
> 
> ```
>   "devDependencies": {
>     "@angular-devkit/build-angular": "0.803.20",
>     "@angular/cli": "8.3.20",
> ```
This worked for me also. I also noticed that removing any modules that referenced "ngx-bootstrap" caused the problem to go away.

**Comment by SkanderEchaoui** [comment]
Hello. It happened to me after upgrading to angular/cli@8.3.21 and angular-devkit/build-angular@0.803.21. Going back to .20 "fixed" the issue.

thnx mate, this worked for me.

**Comment by tmtron** [comment]
- using newest node version (`12.14.0`) didn't work for me
- solution for now is to go back to old dev-deps:
```
"devDependencies": {
    "@angular-devkit/build-angular": "0.803.20",
    "@angular/cli": "8.3.20",
```

- increasing the memory `node --max_old_space_size=12000` also works as a workaround

**Comment by robertrandallphillips** [comment]
Updating node to 12.14.0 worked for me with the latest build-angular and cli packages. Thanks.

**Comment by pavlkara1** [comment]
After updating to 8.3.21, I also have this issue but only for non production issues. When building with `ng build --prod` , node doesn't run out of memory.  (Using node v12.14.0)

**Comment by rocts** [comment]
Same issue for me. Can we get this issue reopened or is there another issue tracking it?  This must be affecting most users.

**Comment by tmtron** [comment]
I guess this is related (or even the same): #16515

**Comment by dicbrus** [comment]
"@angular-devkit/build-angular": "~0.803.21",
"@angular/cli": "^8.3.21",

still happens

**Comment by AhsanAyaz** [comment]
@dicbrus could you try this?
https://github.com/angular/angular-cli/issues/16515#issuecomment-571718620

**Comment by jarekcieslak** [comment]
I accidentally stumbled upon this thread. In my case CircleCI was failing because of displaying too much progress in the build step. A small change by running `ng build --no-progress` helped.
